Handle loss of file watcher functionality when folder being watched is recreated (after deletion).

anandgupta08 opened the pull request:
Handle loss of file watcher functionality when folder being watched is recreated (after deletion).

Also, handle possible sudden burst of new changes on folder re-creation which otherwise leads to exceptionally high number of browser reload requests(within a very short period of few seconds) that makes browser unresponsive.
This is typically helpful when live reloading is set up  for standalone folder of SPA frameworks (like Angular / React/ Vue) but ASP.NET Core server continues to run without restarting, and we want to rebuild (stop and rebuild) the client side code (e.g. Angular / React) in watch mode (e.g. 'ng build --watch' in case of Angular). In such cases the SPA output folder (e.g. 'dist' in Angular or 'public' in React) are typically deleted and recreated as part of its clean build process. In such cases Live Reload fails as the FileWatcher stops watching (at least in Windows) when the folder it is supposed to watch has been deleted and recreated. The only way to start monitoring the newly recreated standalone folder is by restarting the ASP.NET Core service even though the change was in client.
Additionally, possible renaming (either programmatically or manually for any reason in any scenario) of  the folder to be monitored is also handled. For example if we delete a folder manually, the FileWatcher stops watching it (at least in Windows) and creating it again manually is considered renaming in Windows - hence handling renaming of the folder being monitored can make this middleware more robust and reliable.

anandgupta08 committed:
Handle loss of file watcher functionality when folder being watched is recreated (after deletion).
Also, handle possible sudden burst of new changes on folder re-creation which otherwise leads to exceptionally high number of browser reload requests(within a very short period of few seconds) that makes browser unresponsive.
This is typically helpful when live reloading is set up  for standalone folder of SPA frameworks (like Angular / React/ Vue) but ASP.NET Core server continues to run without restarting, and we want to rebuild (stop and rebuild) the client side code (e.g. Angular / React) in watch mode (e.g. 'ng build --watch' in case of Angular). In such cases the SPA output folder (e.g. 'dist' in Angular or 'public' in React) are typically deleted and recreated as part of its clean build process. In such cases Live Reload fails as the FileWatcher stops watching (at least in Windows) when the folder it is supposed to watch has been deleted and recreated. The only way to start monitoring the newly recreated standalone folder is by restarting the ASP.NET Core service even though the change was in client.
Additionally, possible renaming (either programmatically or manually for any reason in any scenario) of  the folder to be monitored is also handled. For example if we delete a folder manually, the FileWatcher stops watching it (at least in Windows) and creating it again manually is considered renaming in Windows - hence handling renaming of the folder being monitored can make this middleware more robust and reliable.

RickStrahl commented:
This looks good and it seems to not break anything.

One question - you have `FolderWatcher.IncludeSubdirectories = false;` which  doesn't seem right here - wouldn't you want to monitor down the subtree for changes?

Also if the folderwatcher is there I think the file watcher can probably remove the   | `NotifyFilters.DirectoryName` in the file watcher and then be more explicit in the the folder watcher 
 with explicit NotifyFilter settings. Offhand I'm not sure what would be needed but Specifically `NotifyFilters.DirectoryName | NotifyFilters.LastWrite`  seems like the right thing?

Think like this perhaps?

```cs
private static void StartFilesWatcher()
{
    FileWatcher = new FileSystemWatcher(FolderToMonitorPath);
    FileWatcher.IncludeSubdirectories = true;
    FileWatcher.Filter = "*.*";
    FileWatcher.NotifyFilter = NotifyFilters.LastWrite
                               | NotifyFilters.FileName;
    FileWatcher.EnableRaisingEvents = true;
    

    FileWatcher.Changed += FileWatcher_Changed;
    FileWatcher.Created += FileWatcher_Changed;
    FileWatcher.Renamed += FileWatcher_Renamed;
}

private static void StartFolderWatcher()
{
    var parentPath = Path.GetDirectoryName(FolderToMonitorPath);
    var folderName = Path.GetFileName(FolderToMonitorPath);

    FolderWatcher = new FileSystemWatcher(parentPath);
    FolderWatcher.IncludeSubdirectories = false;
    FolderWatcher.Filter = folderName;
    FileWatcher.NotifyFilter = NotifyFilters.LastWrite
                               | NotifyFilters.DirectoryName;
    FolderWatcher.EnableRaisingEvents = true;
    
    FolderWatcher.Created += FolderWatcher_Created;
    FolderWatcher.Deleted += FolderWatcher_Deleted;
    FolderWatcher.Renamed += FolderWatcher_Renamed;
}
```

RickStrahl commented:
Ok so I played around with this and I don't see the folder watcher ever firing, no matter what I do. I reset the `NotifyFilter` to its default and still I don't see it firing on either directory creation or deletion. Not in the root folder (ie. the IncludeSubDirectories wouldn't apply) or in a sub-directory which actually fires the File watcher.

I see the **file watcher** firing on created folders, and nothing happening on deleted folders.

It seems to me that the singular watcher could probably handle both folders and directories by checking if the target is a folder or a file perhaps which would simplify this code quite a bit. Or am I missing the point here?

anandgupta08 commented:
@RickStrahl I have created test harness console app for the file watcher to help play around the file creation / editing / deletion and folder creation / renaming / deletion.
https://github.com/anandgupta08/TestHarnessLiveReload.git
I have commented out the code that initiates the refresh using the middleware with Console.WriteLine. Also, I have added Console.WriteLine to the event handlers for Folder_Created, Folder_Deleted and Folder_Renamed, so it is easier to observe and validate the firing of the events.

`
FolderWatcher.IncludeSubdirectories = false;
` 
should be fine because we only want to watch the specific folder specified as FolderToMonitorName . The subdirectories and files inside this folder is being monitored by existing code using ` private static System.IO.FileSystemWatcher FileWatcher;`

I will further play a bit to determine if we can be more specific about `NotifyFilters` for both folder and file watcher.
**In the meantime, if you could play around with the test harness and confirm it works functionally as expected, that would be great.**

Note: 
When you play with the test harnesss, please create / edit / delete files with extensions as per the default `ClientFileExtensions` settings:
` public string ClientFileExtensions { get; set; } = ".cshtml,.css,.js,.htm,.html,.ts,.razor";`

Playing with .txt files will not work (i.e. not fire events we expect) without modifying the default settings for `ClientFileExtensions`

Also, I have modified the `DisposeFilesWatcher` and `DisposeFolderWatcher` a bit  in the TestHarness code (Vs my original pull request) to make it more robust as I noticed in some situations, calling Dispose was not enough and unsubscribing the event handlers is needed. Once you confirm the test harness works as expected, I will checkin the code modifications of Dispose

anandgupta08 committed:
Improved the DisposeFilesWatcher and DisposeFolderWatcher to ensure event subscriptions are unsubscribed. In some renaming situations, just disposing the watcher does not seem to be sufficiient and event handlers fired multiple times.

anandgupta08 commented:
@RickStrahl , Did you get a chance to go through my Test Harness project :
https://github.com/anandgupta08/TestHarnessLiveReload
and confirm the functionality is working ad expected.

I explored the possibility if a single watcher (folder + files), but don't think that will work as effectively / granularly if we want to filter based on specific file extensions (which is the current functionality and my opinion is we must have this flexibility).

RickStrahl commented:
Thank you for this - I haven't had time to take a look at this yet. I hope to get to it tonight or tomorrow.

anandgupta08 commented:
@RickStrahl Did you get a chance to review the PR and look at the test harness project I created at https://github.com/anandgupta08/TestHarnessLiveReload ?

RickStrahl commented:
Finally had a chance to look at this. Took me a little to understand the scenario you're trying to demonstrate, but I agree a good fix.

Before I merge though, can you tell me the use case for this scenario? The assumption appears to be that that the target --webroot folder is either created or deleted. I'm trying to think when this makes sense?

anandgupta08 commented:
Yes, you understood it correctly.

This is typically helpful when live reloading is set up for standalone folder of SPA frameworks (like Angular / React/ Vue) but ASP.NET Core server continues to run without restarting, and we want to rebuild (stop and rebuild) the client side code (e.g. Angular / React) in watch mode (e.g. 'ng build --watch' in case of Angular). In such cases the SPA output folder (e.g. 'dist' in Angular or 'public' in React) are typically deleted and recreated as part of its clean build process. In such cases Live Reload fails as the FileWatcher stops watching (at least in Windows) when the folder it is supposed to watch has been deleted and recreated. The only way to start monitoring the newly recreated standalone folder is by restarting the ASP.NET Core service even though the change was in client.
Additionally, possible renaming (either programmatically or manually for any reason in any scenario) of the folder to be monitored is also handled. For example if we delete a folder manually, the FileWatcher stops watching it (at least in Windows) and creating it again manually is considered renaming in Windows - hence handling renaming of the folder being monitored can make this middleware more robust and reliable.

RickStrahl commented:
Yes I actually ran into this very issue with an Angular app that I build into a .NET Core app wwwroot folder - just today as I was giving a demo... 

Merged. Thank you for this.

anandgupta08 commented:
@RickStrahl Thank you for accepting my PR. Glad, you found the PR useful.
